### PR TITLE
Moving default backend to Pulsar

### PR DESCRIPTION
**Pull request by manquius**

### CODE REVIEW

File: consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java [hunk 1]
```diff
@@ -1,14 +1,31 @@
...
+            adapter = getClientByName(clientString);
+        }
+        return adapter;
+    }
+
+    private ConsumeClient getClientByName(String clientString) throws ClientCreationException {
+        if(clientString.equalsIgnoreCase("KAFKA")) {
+            adapter = new KafkaConsumeClient();
+        }
+        if(clientString.equalsIgnoreCase("REDIS")) {
```
Comment (rliberal) [inline consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java]: Create an enum for this.

File: consumer/src/main/java/com/manquius/twelvefactor/client/PulsarConsumeClient.java [hunk 2]
```diff
@@ -5,14 +5,18 @@
...
+
 
 class PulsarConsumeClient implements ConsumeClient {
 
     private final PulsarClient client;
     private final ConsumerBuilder<String> builder;
 
     public PulsarConsumeClient() throws ClientCreationException {
-        String pulsarBrokerRootUrl = "pulsar://localhost:6650";
+        String host = ofNullable(System.getenv("TWELVEFACTOR_PULSAR_PROXY_SERVICE_HOST")).orElse("localhost");
```
Comment (rliberal) [inline consumer/src/main/java/com/manquius/twelvefactor/client/PulsarConsumeClient.java:17]: Use config map here as env variable

File: consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java (hunk 1 above)
Comment (rliberal) [reply consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java]: In order to have a list of all your clients.

File: consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java [hunk 3]
```diff
@@ -1,14 +1,31 @@
...
 
-    private ConsumeClient redisConsumeClient;
+public class ConsumeClientProvider {
 
+    private ConsumeClient adapter;
 
     public ConsumeClient getClient() throws ClientCreationException {
-        if(redisConsumeClient == null) {
-            redisConsumeClient = new KafkaConsumeClient();
+        if (null == adapter) {
```
Comment (rliberal) [inline consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java:10]: Check these. 

https://www.baeldung.com/java-avoid-null-check
https://codeburst.io/better-null-checking-in-java-369f1701155f


Let me know if we can try some of this for null.

File: consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java [hunk 4]
```diff
@@ -1,14 +1,31 @@
...
+            final String clientString = ofNullable(System.getenv("CONSUME_CLIENT")).orElse("PULSAR");
+            adapter = getClientByName(clientString);
+        }
+        return adapter;
+    }
+
+    private ConsumeClient getClientByName(String clientString) throws ClientCreationException {
```
Comment (rliberal) [inline consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java]: final for client string

File: consumer/src/main/java/com/manquius/twelvefactor/client/PulsarConsumeClient.java [hunk 5]
```diff
@@ -24,7 +28,7 @@ public PulsarConsumeClient() throws ClientCreationException {
     @Override
     public List<String> consume(String topic) throws ConsumeException {
```
Comment (rliberal) [inline consumer/src/main/java/com/manquius/twelvefactor/client/PulsarConsumeClient.java]: use final again

**Review by rliberal — CHANGES_REQUESTED** [review]

File: consumer/src/main/java/com/manquius/twelvefactor/client/PulsarConsumeClient.java (hunk 2 above)
Comment (manquius) [reply consumer/src/main/java/com/manquius/twelvefactor/client/PulsarConsumeClient.java:17]: The used variable is automatically provided by kubernetes to the different containers.

File: consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java (hunk 3 above)
Comment (manquius) [reply consumer/src/main/java/com/manquius/twelvefactor/client/ConsumeClientProvider.java:10]: This is a one time initialization, proposed changes are meaningless in this context.

**Review by rliberal — APPROVED** [review]
Sounds Good.

**Pull request merged**